### DISCUSSION TITLE
How to add a new strategy?

### DISCUSSION THREAD

**Discussion by umutt** [EA31337/EA31337#305] [Q&A] [answered]
Hi, is there a guide for adding new custom strategies and indicators ?
For example if i have a custom indicator like [SuperTrend](https://www.mql5.com/en/code/22611), how can i convert and import it into the system.
Thanks.

**Comment by kenorb** [comment, marked as answer]
There is no guide yet, as the steps are not finalized. In general, you should check the structure of other [strategies](https://github.com/EA31337/EA31337-strategies).

In general, to create a new strategy, you can:

1. Create a new repository by forking one of the [existing strategies](https://github.com/topics/ea31337-strategy), or use them as a template (_Use this template_ option). For the basic strategy (based on 1 indicator), you can use [Demo strategy](https://github.com/EA31337/Strategy-Demo) as a starting point.
2. Replace 'Demo' (code and filenames) with your new strategy name.
3. In `Stg_X.mqh` (strategy class), find a line with `SetIndicator()` and point it to your existing indicator class. If your indicator is new and needs to be added, check the separate steps below.

Adding a new indicator to strategy:

1. If your indicator is new (not part of framework), upload indicator file to new `Indicators/` directory of your strategy repository. See [example here](https://github.com/EA31337/Strategy-TMA_True/tree/v1.007/Indicators).
2. Optionally, if you're uploading MQL4 indicator, you can create a MQL5 wrapper code around it by creating [`.mq5` file such as this one](https://github.com/EA31337/Strategy-TMA_True/blob/v1.007/Indicators/TMA_True.mq5).
3. Create an indicator class to implement your indicator file (`Indi_YourIndicator.mqh`) and change the code accordingly. See [`Indi_TMA_True.mqh`](https://github.com/EA31337/Strategy-TMA_True/blob/v1.007/Indi_TMA_True.mqh) for example.
4. Compile a strategy file to check for any errors. Comment out any lines of code that you're not sure of (can be changed later on).

Updating strategy class:

1. If your indicator provides any input parameters, these should be added to strategy class (e.g. `StrategyName_Indi_IndicatorName_Shift` for indicator shift).
2. Any new parameters should be passed in structure with default indicator values. See: [Stg_TMA_True_Indi_TMA_True_Params_Defaults](https://github.com/EA31337/Strategy-TMA_True/blob/v1.007/Stg_TMA_True.mqh#L54-L59) for the example.
3. Writing trading conditions: Edit `Stg_YourStrategy.mqh` and update `SignalOpen()` code to put new signal conditions. See: [Stg_TMA_True.mqh](https://github.com/EA31337/Strategy-TMA_True/blob/v1.007/Stg_TMA_True.mqh) for the example.

Testing new strategy:

1. Each strategy should come with files that you can compile and run (`.mq5`/`.mq4`). This can be used to test your strategy alone without being part of EA.
2. Optimize any params that you feel should work better in general (default recommended symbol is: EURUSD).

Proposing strategy to be part of EA:

1. Once your strategy works as expected, you can create a new request on [EA31337-strategies](https://github.com/EA31337/EA31337-strategies/issues) to be included in EA (so it's going to be forked and included into the main collection).
2. Or you can include it on your own, by updating the code in relevant parts (enums, functions) and re-compiling the EA ([EA31337-Libre](https://github.com/EA31337/EA31337-Libre) is a good start).

Having said above, I'll try to help and implement this mentioned indicator and let you know how it goes.

----

## Demo

Update: I've already generated the initial strategy. And I'm attaching a 3 minute quick textual video how the strategy has been created:

[![asciicast](https://asciinema.org/a/439243.svg)](https://asciinema.org/a/439243)

It's a bit code mess, but each version is going to simplify the process.

Hope that helps!

**Reply by umutt** [reply]
Thank you so much for your detailed explanation. At this point, I'm trying to understand the framework and you are really helpful. 
Looking forward to follow supertrend integration process. I hope some day I will propose my favorite trend tracker strategies like https://www.tradingview.com/script/zVhoDQME/ to EA31337 community.

**Reply by kenorb** [reply]
I've added your strategy idea/suggestion to new section at [Strategy Ideas](https://github.com/EA31337/EA31337-strategies/discussions/categories/strategy-ideas).

**Reply by kenorb** [reply]
I've uploaded a textual video above to show the conversion process from Demo to any strategy (as example). It's just initial code which compiles and runs, there is still some work around integration and finding the right signals for this new strategy.

**Reply by umutt** [reply]
Thanks, I have managed to compile and run with following little modification, I guess it is something about master vs dev branch.

    ADD**_stg_params.SetIndicator(new Indi_SuperTrend(_indi_params));**
    ChartParams _cparams(_tf, _Symbol);
    TradeParams _tparams;
    Strategy *_strat = new Stg_SuperTrend(_stg_params, _tparams, _cparams, "SuperTrend");
    COMMENT**//_strat.SetIndicator(new Indi_SuperTrend(_indi_params));**
